src/utils/lazy.ts

```diff
@@ -7,21 +7,68 @@ export class Lazy<T> {
     private _isValueCreated: boolean = false;
     private _value: T | undefined;
 
-    public constructor(private readonly valueFactory: () => T) {
+    public constructor(private readonly valueFactory: () => T, private readonly _valueLifetime?: number) {
     }
 
     public get isValueCreated(): boolean {
         return this._isValueCreated;
     }
 
     public get value(): T {
-        if (!this._isValueCreated) {
-            this._value = this.valueFactory();
-            this._isValueCreated = true;
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._value = this.valueFactory();
+        this._isValueCreated = true;
+
+        if (this._valueLifetime) {
+            const reset = setTimeout(() => {
+                this._isValueCreated = false;
+                this._value = undefined;
+                clearTimeout(reset);
+            }, this._valueLifetime);
         }
 
         return this._value;
     }
 }
 
-export default Lazy;
+export class AsyncLazy<T> {
+    private _isValueCreated: boolean = false;
+    private _value: T | undefined;
+    private _valuePromise: Promise<T> | undefined;
+
+    public constructor(private readonly valueFactory: () => Promise<T>, private readonly _valueLifetime?: number) {
+    }
+
+    public async getValue(): Promise<T> {
```

Comment:
considering the complexity of this class, please add unit test for this.

Reply:
Absolutely!

Reply:
I added UTs, ran them ~10 times locally without any failures.

Reply:
Great. Thanks!